examples_tests/0.ImportanceSamplingEnvMaps/main.cpp

```diff
@@ -0,0 +1,668 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#define _NBL_STATIC_LIB_
+#include <nabla.h>
+#include "../../../include/nbl/asset/filters/CSummedAreaTableImageFilter.h"
+#include "nbl/ext/FullScreenTriangle/FullScreenTriangle.h"
+#include "nbl/ext/ScreenShot/ScreenShot.h"
+#include "glm/glm.hpp"
+#include "../common/QToQuitEventReceiver.h"
+
+using namespace nbl;
+using namespace asset;
+using namespace core;
+using namespace video;
+
+using SATFilter = CSummedAreaTableImageFilter<false>;
+
+static core::smart_refctd_ptr<ICPUBuffer> computeLuminancePdf(smart_refctd_ptr<ICPUImage> envmap)
+{
+	const core::vector2d<uint32_t> envmapExtent = { envmap->getCreationParameters().extent.width, envmap->getCreationParameters().extent.height };
+	const uint32_t channelCount = getFormatChannelCount(envmap->getCreationParameters().format);
+	
+	const core::vector2d<uint32_t> pdfDomainExtent = { envmapExtent.X, envmapExtent.Y };
+
+	const size_t outBufferSize = pdfDomainExtent.X * pdfDomainExtent.Y * sizeof(double);
+	core::smart_refctd_ptr<ICPUBuffer> outBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(outBufferSize);
+
+	const double luminanceScales[4] = { 0.2126729 , 0.7151522, 0.0721750, 0.0 };
+
+	float* envmapPixel = (float*)envmap->getBuffer()->getPointer();
+	double* outPixel = (double*)outBuffer->getPointer();
+
+	for (uint32_t y = 0; y < pdfDomainExtent.Y; ++y)
+	{
+		const double sinTheta = core::sin(core::PI<double>() * ((y + 0.5) / (double)pdfDomainExtent.Y));
+
+		for (uint32_t x = 0; x < pdfDomainExtent.X; ++x)
+		{
+			double result = 0.0;
+			for (uint32_t ch = 0; ch < channelCount; ++ch)
+				result += luminanceScales[ch] * envmapPixel[ch];
+
+			*outPixel++ = result * sinTheta;
+			envmapPixel += channelCount;
+		}
+	}
+
+	return outBuffer;
+}
+
+// Returns the offset into the passed array the element at which is <= the passed element (`x`)
+// returns offset = -1 if passed element is < the element at index 0
+static int32_t bisectionSearch(const double* arr, const uint32_t arrCount, const double x, double* xFound)
+{
+	int32_t offset = std::upper_bound(arr, arr + arrCount, x) - arr - 1u;
+	double dx = 0.0;
+	if (offset == -1)
+		dx = x / arr[offset + 1];
+	else
+		dx = (x - arr[offset]) / (arr[offset + 1] - arr[offset]);
+
+	// This assumes array values to be in the range [0,1) which is fine for our purposes because we use them as texture coordinates
+	if (xFound)
+		*xFound = (offset + 1 + dx) / arrCount;
```

Comment:
offset+1 is the location of the value that is greater than `x` , and then you're adding `dx` to it... this seems wrong

shouldn't it be `(offset+dx)` ?

Reply:
@devshgraphicsprogramming It was misleading me at first, but notice that @achalpandeyy subtracts -1 from upper_bound returned offset so here actually the offset is actually refering the value before that which is less than `x`, I think he's just making up for that `-1` a few lines back.

Reply:
thanks